File: RailgunVB/Commands/Utilities/WhoIs.vb
Imports System.Text
Imports Discord
Imports Discord.Commands
Imports RailgunVB.Core

Namespace Commands.Utilities
    
    <Group("whois")>
    Public Class WhoIs
        Inherits SystemBase
        
        <Command>
        Public Async Function WhoisAsync(userId As ULong) As Task
            Dim user As IUser = Await Context.Client.GetUserAsync(userId)
            
            If user Is Nothing
                Await ReplyAsync("Can not find user.")
                Return
            End If
            
            Await WhoisAsync(user)
        End Function
        
        <Command>
        Public Async Function WhoisAsync(Optional pUser As IUser = Nothing) As Task
            Dim user As IUser = If(pUser Is Nothing, Context.User, pUser)
            Dim username As String = $"{user.Username}#{user.DiscriminatorValue}"
            Dim isHuman As String = If(user.IsBot Or user.IsWebhook, "Bot", "User")
            Dim guilds As IReadOnlyCollection(Of IGuild) = Await Context.Client.GetGuildsAsync()
            Dim locatedOn = 0
            Dim builder As New EmbedBuilder With {
                    .Title = $"Who is {Format.Bold(username)}?",
                    .Color = Color.Blue,
                    .ThumbnailUrl = user.GetAvatarUrl(),
                    .Footer = new EmbedFooterBuilder() With {
                        .Text = $"User ID : {user.Id}"
                    }
                }
            
            For Each guild As IGuild In guilds
                If (Await guild.GetUserAsync(user.Id)) IsNot Nothing Then locatedOn += 1
            Next
            
            builder.AddField("User Type:", isHuman, True) _
                .AddField("Registered:", user.CreatedAt, True) _
                .AddField("Located On:", locatedOn, True)
            
            Dim gUser As IGuildUser = Await Context.Guild.GetUserAsync(user.Id)
            
            If gUser IsNot Nothing
                Dim roles As New StringBuilder
                
                For Each roleId As ULong In gUser.RoleIds
                    roles.AppendFormat("| {0} |", Context.Guild.GetRole(roleId).Mention)
                Next
                
                builder.AddField("Server Nickname:", If(gUser.Nickname, "N/A"), true) _
                    .AddField("Joined Server At:", If(gUser.JoinedAt, "UNKNOWN"), true) _
                    .AddField("Current Server Roles:", If(roles.ToString(), "N/A")) _
                    .AddField("ESPer Level:", GetEsperLevel(gUser))
            End If
            
            Await ReplyAsync("", embed := builder.Build())
        End Function
        
        Private Function GetEsperLevel(gUser As IGuildUser) As Integer
            If Context.Guild.OwnerId = gUser.Id Then Return 6
            
            Dim score = 0
            
            For Each perm As GuildPermission In gUser.GuildPermissions.ToList()
                Select perm
                    Case GuildPermission.Administrator
                        Return 5
                    Case GuildPermission.ManageChannels, GuildPermission.ManageEmojis, GuildPermission.ManageGuild, 
                        GuildPermission.ManageRoles, GuildPermission.ManageWebhooks, GuildPermission.ViewAuditLog,
                        GuildPermission.ManageNicknames, GuildPermission.ManageMessages
                        score += 10
                        Exit Select
                    Case GuildPermission.BanMembers, GuildPermission.DeafenMembers, GuildPermission.KickMembers, 
                        GuildPermission.MoveMembers, GuildPermission.MuteMembers, GuildPermission.MentionEveryone, 
                        GuildPermission.SendTTSMessages, GuildPermission.PrioritySpeaker
                        score += 5
                        Exit Select
                    Case GuildPermission.UseVAD, GuildPermission.UseExternalEmojis, GuildPermission.AttachFiles, 
                        GuildPermission.CreateInstantInvite, GuildPermission.AddReactions, GuildPermission.ChangeNickname,
                        GuildPermission.EmbedLinks, GuildPermission.ReadMessageHistory
                        score += 2
                        Exit Select
                    Case GuildPermission.SendMessages, GuildPermission.ViewChannel, GuildPermission.Connect, 
                        GuildPermission.Speak
                        score += 1
                        Exit Select
                End Select
            Next
            
            Dim percent As Double = Math.Round((score / 140) * 100)
            
            Select percent
                Case Is < 10
                    Return 0
                Case Is < 20
                    Return 1
                Case Is < 40
                    Return 2
                Case Is < 60
                    Return 3
                Case Else
                    Return 4
            End Select
        End Function
    
    End Class
    
End NameSpace

File: Railgun/Commands/Utilities/WhoIs.cs
using System;
using System.Text;
using System.Threading.Tasks;
using Discord;
using Finite.Commands;
using Railgun.Core.Commands;

namespace Railgun.Commands.Utilities
{
    [Alias("whois")]
    public class WhoIs : SystemBase
    {
        private int GetEsperLevel(IGuildUser user) {
            if (Context.Guild.OwnerId == user.Id) return 6;

            var score = 0;

            foreach (var perm in user.GuildPermissions.ToList()) {
                switch (perm) {
                    case GuildPermission.Administrator:
                        return 5;
                    case GuildPermission.ManageChannels:
                    case GuildPermission.ManageEmojis:
                    case GuildPermission.ManageGuild:
                    case GuildPermission.ManageRoles:
                    case GuildPermission.ManageWebhooks:
                    case GuildPermission.ViewAuditLog:
                    case GuildPermission.ManageNicknames:
                    case GuildPermission.ManageMessages:
                        score += 10;
                        break;
                    case GuildPermission.BanMembers:
                    case GuildPermission.DeafenMembers:
                    case GuildPermission.KickMembers:
                    case GuildPermission.MoveMembers:
                    case GuildPermission.MuteMembers:
                    case GuildPermission.MentionEveryone:
                    case GuildPermission.SendTTSMessages:
                    case GuildPermission.PrioritySpeaker:
                        score += 5;
                        break;
                    case GuildPermission.UseVAD:
                    case GuildPermission.UseExternalEmojis:
                    case GuildPermission.AttachFiles:
                    case GuildPermission.CreateInstantInvite:
                    case GuildPermission.AddReactions:
                    case GuildPermission.ChangeNickname:
                    case GuildPermission.EmbedLinks:
                    case GuildPermission.ReadMessageHistory:
                        score += 2;
                        break;
                    default:
                        score += 1;
                        break;
                }
            }

            var percent = Math.Round((score / 140) * 100.00);

            if (percent < 10) return 0;
            else if (percent < 20) return 1;
            else if (percent < 40) return 2;
            else if (percent < 60) return 3;
            else return 4;
        }

        [Command]
        public Task WhoIsAsync() => WhoIsAsync(Context.Author);

        [Command]
        public async Task WhoIsAsync(ulong userId) {
            var user = await Context.Client.GetUserAsync(userId);

            if (user == null) {
                await ReplyAsync("Can not find user.");
                return;
            }

            await WhoIsAsync(user);
        }

        [Command]
        public async Task WhoIsAsync(IUser user) {
            var username = $"{user.Username}#{user.DiscriminatorValue}";
            var isHuman = user.IsBot | user.IsWebhook ? "Bot" : "User";
            var guilds = await Context.Client.GetGuildsAsync();
            var locatedOn = 0;

            var embedBuilder = new EmbedBuilder() {
                Title = $"Who is {Format.Bold(username)}?",
                Color = Color.Blue,
                ThumbnailUrl = user.GetAvatarUrl(),
                Footer = new EmbedFooterBuilder() { Text = $"User ID : {user.Id}" }
            };

            foreach (var guild in guilds) if ((await guild.GetUserAsync(user.Id)) != null) locatedOn++;

            embedBuilder.AddField("User Type:", isHuman, true)
                .AddField("Registered:", user.CreatedAt, true)
                .AddField("Located On:", $"{locatedOn} servers", true);
            
            var gUser = await Context.Guild.GetUserAsync(user.Id);

            if (gUser != null) {
                var roles = new StringBuilder();

                foreach (var roleId in gUser.RoleIds) roles.AppendFormat("{0} ", Context.Guild.GetRole(roleId).Mention);

                embedBuilder.AddField("Server Nickname:", gUser.Nickname ?? Format.Bold("N/A"), true)
                    .AddField("Joined Server At:", gUser.JoinedAt.ToString() ?? Format.Bold("Recently"), true)
                    .AddField("Current Server Roles:", roles.ToString() ?? Format.Bold("N/A"))
                    .AddField("ESPer Level:", GetEsperLevel(gUser));
            }

            await ReplyAsync(embed:embedBuilder.Build());
        }
    }
}